tutorials/multi/1_graph_classification.py

```diff
@@ -0,0 +1,241 @@
+# To ensure same initial model parameters across processes, we need to set a
+# same random seed before model initialization. Once constructed, we wrap the
+# model instance with :func:`~torch.nn.parallel.DistributedDataParallel`.
+#
+
+import torch
+from torch.nn.parallel import DistributedDataParallel
+
+def init_model(seed, device):
+    torch.manual_seed(seed)
+    model = GIN().to(device)
+    model = DistributedDataParallel(model, device_ids=[device], output_device=device)
+
+    return model
+
+###############################################################################
+# Main Function for One Process
+# -----------------------------
+#
+# Define the model evaluation function as in the single-GPU setting.
+#
+
+def evaluate(model, dataloader, device):
+    model.eval()
+
+    total = 0
+    total_correct = 0
+
+    for bg, labels in dataloader:
+        bg = bg.to(device)
+        labels = labels.to(device)
+        # Get input node features
+        feats = bg.ndata.pop('attr')
+        with torch.no_grad():
+            pred = model(bg, feats)
+        _, pred = torch.max(pred, 1)
+        total += len(labels)
+        total_correct += (pred == labels).sum().cpu().item()
+
+    return 1.0 * total_correct / total
+
+###############################################################################
+# Define the main function for one process. Since different processes handle
+# different data batches, some of them can finish their task earlier.
+# `dist.barrier()` forces faster processes to wait until all processes
+# have encountered the barrier.
+#
+
+from torch.optim import Adam
+
+def main(rank, world_size, dataset, seed=0):
+    init_process_group(world_size, rank)
+    # Assume the GPU ID to be the same as the process ID
+    device = torch.device('cuda:{:d}'.format(rank))
+    torch.cuda.set_device(device)
+
+    model = init_model(seed, device)
+    criterion = nn.CrossEntropyLoss()
+    optimizer = Adam(model.parameters(), lr=0.01)
+
+    train_loader, val_loader, test_loader = get_dataloaders(dataset,
+                                                            seed,
+                                                            world_size,
+                                                            rank)
+    for epoch in range(5):
+        model.train()
+        # The line below ensures all processes use a different
+        # random ordering in data loading for each epoch.
+        train_loader.set_epoch(epoch)
+
+        total_loss = 0
+        for bg, labels in train_loader:
+            bg = bg.to(device)
+            labels = labels.to(device)
+            feats = bg.ndata.pop('attr')
+            pred = model(bg, feats)
+
+            loss = criterion(pred, labels)
+            total_loss += loss.cpu().item()
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+        loss = total_loss
+        print('Loss: {:.4f}'.format(loss))
+        dist.barrier()
```

Comment:
In most cases, it should work fine without the barrier.

Reply:
Then, what are the cases that one needs to add it? My concern would be that the readers are all gonna follow it, and might not be the best practice as it is not necessary.

Reply:
How about simply removing them? As you said, they are unnecessary in this example. To fully explain when this will be necessary might be out of the scope of this tutorial.

Reply:
OK let's remove it.